fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u opened the pull request:
# Description
Improve SAML command to stop wrong warnings. Tests updated too.

# Private Ticket 
https://2u-internal.atlassian.net/jira/software/c/projects/BOMS/boards/3017?assignee=712020%3A61c35560-3472-4e12-b833-884e5c4bbff4&selectedIssue=BOMS-64

# Related PR
https://github.com/openedx/edx-platform/pull/37330

Saml Configuration Report 
```
SAML Configuration Check Report
==================================================

[WARNING] Provider (id=74, name=, slug=default-apphelix, site_id=1) has outdated SAML config (id=50) which should be updated to the current SAML config (id=77).
[INFO] Provider (id=74, name=, slug=default-apphelix, site_id=1) has SAML config (id=50, slug='default-mock') that does not match the provider's slug.
[WARNING] Provider (id=77, name=, slug=default-melody, site_id=1, enabled=False) has outdated SAML config (id=50) which should be updated to the current SAML config (id=77).
[INFO] Provider (id=77, name=, slug=default-melody, site_id=1, enabled=False) has SAML config (id=50, slug='default-mock') that does not match the provider's slug.
[WARNING] Provider (id=78, name=, slug=default-chintan, site_id=1, enabled=False) has outdated SAML config (id=50) which should be updated to the current SAML config (id=77).
[INFO] Provider (id=78, name=, slug=default-chintan, site_id=1, enabled=False) has SAML config (id=50, slug='default-mock') that does not match the provider's slug.
[INFO] Provider (id=88, name=, slug=vaibhav1, site_id=1) has SAML config (id=61, slug='apphelix1') that does not match the provider's slug.
[INFO] Provider (id=89, name=, slug=vaibhav2, site_id=1) has SAML config (id=61, slug='apphelix1') that does not match the provider's slug.
[WARNING] Provider (id=117, name=, slug=default-slug, site_id=6, enabled=False) has SAML config (id=73, enabled=False).
[WARNING] Provider (id=117, name=, slug=default-slug, site_id=6, enabled=False) SAML config (id=73, site_id=1) does not match the provider's site_id.
[INFO] Provider (id=117, name=, slug=default-slug, site_id=6, enabled=False) has SAML config (id=73, slug='slug', enabled=False) that does not match the provider's slug.
[WARNING] Provider (id=119, name=, slug=slug, site_id=1, enabled=False) has SAML config (id=76, enabled=False).
[WARNING] Provider (id=119, name=, slug=slug, site_id=1, enabled=False) SAML config (id=76, site_id=8) does not match the provider's site_id.
[WARNING] Provider (id=123, name=, slug=default-edx, site_id=1, enabled=False) has SAML config (id=79, enabled=False).
[INFO] Provider (id=129, name=, slug=defaultssss, site_id=1) has SAML config (id=82, slug='default-s') that does not match the provider's slug.
[INFO] Provider (id=132, name=, slug=diffrent-slug, site_id=4) has SAML config (id=84, slug='test-config') that does not match the provider's slug.
[INFO] Provider (id=137, name=, slug=default0ssd, site_id=1) has SAML config (id=90, slug='default-app') that does not match the provider's slug.
[WARNING] Provider (id=143, name=, slug=pip, site_id=9, enabled=False) has no direct SAML configuration and no matching default configuration was found.
[WARNING] Provider (id=144, name=Provider I, slug=myslug, site_id=10) has no direct SAML configuration and no matching default configuration was found.
[WARNING] Provider (id=145, name=Provider 2, slug=mysecondslug, site_id=10) has no direct SAML configuration and no matching default configuration was found.
[WARNING] Provider (id=146, name=Provider 3, slug=myslug2, site_id=11) has no direct SAML configuration and no matching default configuration was found.
CHECK SUMMARY:
  Providers checked: 22

Informational only:
  Slug mismatches: 9
  Missing configs: 4

Issues requiring attention:
  Outdated: 3
  Site mismatches: 2
  Disabled configs: 3
  Errors: 0

Total issues requiring attention: 8
```

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u commented:
why i change get_config()

when use get_config(), it show many error, same error again again for missing metadata.
real config warning lost in many “No SAMLProviderData found” message.
this kind of Error: `2025-09-23 09:49:45,793 ERROR 339 [common.djangoapps.third_party_auth.models] [user None] [ip None] models.py:873 - No SAMLProviderData found for provider "" with entity id "" and IdP slug "test-config". Run "manage.py saml pull" to fix or debug.
 `
It was counting metadata missing as error, but it not real config problem.
this  was checking config and metadata both. But this command should only check config. Metadata is for --pull command. this was the mix thing which making me confused.

So I have improved it to new logic
I am doing same check, do same config check like get_config(), but no spam error message.
It check direct config and default config like get_config() do. It only say warning if both direct and default config missing.
First check direct, then default, Now config check and metadata check not getting mix

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -112,50 +111,55 @@ def _check_provider_configurations(self):
-                # Check for outdated configuration references
-                if current_config:
-                    if current_config.id != provider_config.saml_configuration_id:
+                    if current_config and current_config.id != provider_config.saml_configuration_id:
```
```suggestion
                    if current_config and (current_config.id != provider_config.saml_configuration_id):
```

jcapphelix replied:
@ktyagiapphelix2u Can you please check this ?

ktyagiapphelix2u replied:
I have implemented this

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -112,50 +111,55 @@ def _check_provider_configurations(self):
-                saml_configuration_slug = provider_config.saml_configuration.slug
-                provider_config_slug = provider_config.slug
+                    saml_configuration_slug = provider_config.saml_configuration.slug
+                    provider_config_slug = provider_config.slug
```
Do they need to be outside if ? 
If yes, can they be written elsewhere ?

ktyagiapphelix2u replied:
Yes i have updated as there are used only once

jcapphelix replied:
It was fine to declare them as vars even if they are only used once. I just meant they needed to be declared in some other place. 

No worries, the new piece of code that you've written to compare them is also okay.

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -112,50 +111,55 @@ def _check_provider_configurations(self):
+                        if not default_config:
+                            self.stdout.write(
+                                f"[WARNING] {provider_info} has no SAML configuration and "
+                                "no matching default configuration was found."
```
Code here checks for SAMLConfigurartion, but the warning is that provider has not SAML config and no matching default config was found. 

Can you explain how that works out ?

ktyagiapphelix2u replied:
The code first checks if provider_config.saml_configuration is None (in the else branch), Then it tries to find a default config using SAMLConfiguration.current(provider_config.site_id, 'default'), If no default config is found (if not default_config), it increments null_config_count, But there's a duplicate warning message that appears in both the if not default_config branch and the except SAMLConfiguration.DoesNotExist branch So I will Fix that

ktyagiapphelix2u replied:
my new logic will check if a provider has no direct config (provider_config.saml_configuration is None), we try to find a default config, If a default config exists, we show an INFO message (new), If no default config exists or an exception occurs, we show appropriate warnings

jcapphelix replied:
This can be resolved.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -112,50 +111,58 @@ def _check_provider_configurations(self):
-                    self.stdout.write(
-                        f"[WARNING] {provider_info} "
-                        f"SAML config (id={provider_config.saml_configuration_id}, slug='{saml_configuration_slug}') "
-                        "does not match the provider's slug."
```
Don't you still want `"does not match the provider's slug."`?

ktyagiapphelix2u replied:
as that message was misleading so i though of removing that statement but as an information we can add it again

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -166,7 +173,7 @@ def _check_provider_configurations(self):
             'outdated_count': {'count': outdated_count, 'requires_attention': True},
             'site_mismatch_count': {'count': site_mismatch_count, 'requires_attention': True},
             'slug_mismatch_count': {'count': slug_mismatch_count, 'requires_attention': True},
```
1. Since this is now an `INFO`, should it no longer require attention? 
2. Should the count output move to the `"CHECK SUMMARY:"` section as well? 
```suggestion
            'slug_mismatch_count': {'count': slug_mismatch_count, 'requires_attention': False},
```

ktyagiapphelix2u replied:
since that is treated as information so i have made the required changes and moved the count output in check_summary.

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -89,7 +84,11 @@ def _handle_run_checks(self):
+        Check each provider configuration for potential issues:
+        - Outdated configuration references
+        - Site ID mismatches
+        - Slug mismatches
```
Should this even be listed, since it just is an inform?

ktyagiapphelix2u replied:
yes i have removed that from the list in addition of adding a statement "Also reports informational data such as slug mismatches."

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -112,50 +111,58 @@ def _check_provider_configurations(self):
+                else:
+                    # Provider has no direct SAML configuration - check for a default one
+                    try:
+                        default_config = SAMLConfiguration.current(provider_config.site_id, 'default')
```
The PR description mentions `get_config()`, but I don't see it in the implementation, and I don't see `null_config_count` > 0 in the test output.

ktyagiapphelix2u replied:
https://github.com/openedx/edx-platform/pull/37377#issuecomment-3326660767

here i explained why i changed from get config to some better approch which is doing the same thing.

ktyagiapphelix2u commented:
The Problem with get_config():

get_config() was mixing configuration validation with metadata validation. It was generating spam error messages about missing metadata ("No SAMLProviderData found"). It was checking both config AND metadata, but this command should only check config. Metadata checking belongs in the --pull command, not the config check command

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented:
Proposed summary output:
```
CHECK SUMMARY:
  Providers checked: 17

Informational only:
  Slug mismatches: 10
  Missing configs: 0

Issues requiring attention:
  Outdated: 3
  Site mismatches: 2
  Errors: 0

Total issues requiring attention: 5
```

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u commented:
@robrap I have updated the test cases and updated the saml config report for missing config one and kept it as warning only and added test case when it will give missing config 
```

Create the provider
Add a new provider:
site: select lost.com
slug: enter my-slug
SAML configuration: leave blank (do not select anything)
Save.
```

```

Check SAML configurations

Filter by site lost.com.
Make sure there is NO SAMLConfiguration with [slug='default' for lost.com.
If one exists, either delete it or change its slug to something else (not default)
```

this will give missing config as `[WARNING] Provider (id=146, name=Provider 3, slug=myslug2, site_id=11) has no direct SAML configuration and no matching default configuration was found.` message.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,52 +111,75 @@ def _check_provider_configurations(self):
-                f"slug={provider_config.slug}, site_id={provider_config.site_id})"
+                f"{disabled_status}Provider (id={provider_config.id}, "
+                f"name={provider_config.name}, slug={provider_config.slug}, "
+                f"site_id={provider_config.site_id})"
```
[optional] I'm wondering about a possible different format:
```
disabled_status = ", enabled=False" if provider_disabled else ""
```
And then:
```suggestion
                f"Provider (id={provider_config.id}, "
                f"name={provider_config.name}, slug={provider_config.slug}, "
                f"site_id={provider_config.site_id}{disabled_status})"
```
The same could be used for disabled saml configs. 

Which do you prefer?

ktyagiapphelix2u replied:
I prefer the suggested format with [enabled=False] at the end because
 representation patterns like `Provider(id=1, name=test, enabled=False)`


as that DISABLED prefix was more of a visual flag

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -89,14 +84,23 @@ def _handle_run_checks(self):
         null_config_count = 0
         error_count = 0
+        disabled_provider_count = 0
+        disabled_config_count = 0
```
I don't think we want additional counts or messages. I think we just want to let people know when the provider and/or sam config is disabled in any of the other messages.

ktyagiapphelix2u replied:
sure we can remove that

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -177,6 +206,45 @@ def _check_provider_configurations(self):
+                self.stdout.write(
+                    f"[INFO] {provider_info} has no direct SAML configuration but "
+                    f"is using default configuration (id={default_config.id})."
+                )
```
I think this should be removed. We are trying to report issues.

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -177,6 +206,45 @@ def _check_provider_configurations(self):
+                return disabled_config_count, null_config_count
+
+            if not default_config.enabled:
+                disabled_config_count += 1
```
This seems like a null config, with just a slightly different issue around the default.
```suggestion
                null_config_count += 1
```

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -177,6 +206,45 @@ def _check_provider_configurations(self):
+                f"[WARNING] {provider_info} has no direct SAML configuration and "
+                "no matching default configuration was found (DoesNotExist)."
+            )
+            null_config_count += 1
```
This isn't in your test output. Is this code reachable, or should it be removed?

ktyagiapphelix2u replied:
Yes, this code is reachable and is covered by tests. The except SAMLConfiguration.DoesNotExist block handles cases where the current() method raises an exception instead of returning None, and there's a specific test that check this scenario.

robrap replied:
A test where this error is mocked does not make this code reachable. The script already handles errors, so can we just drop this try/except?

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -348,24 +349,30 @@ def test_run_checks_outdated_configs(self, mock_set_custom_attribute):
-            output
+        outdated_msg = (
+            f'has outdated SAML config (id={old_config.id}) which should be updated to '
+            f'the current SAML config (id={new_config.id})'
```
Could we add `DISABLED` or `, enabled=False` for each of these that is disabled?

ktyagiapphelix2u replied:
enable=FALSE looks more clear to me for future perspective

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,52 +111,75 @@ def _check_provider_configurations(self):
+            if provider_disabled:
+                disabled_provider_count += 1
+                # Resolution: Enable the provider in Django admin if it should be active
+                self.stdout.write(f"[INFO] {provider_info} is disabled.")
```
Let's just remove this.

ktyagiapphelix2u replied:
yes removed

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,52 +111,75 @@ def _check_provider_configurations(self):
-                        f"SAML config (id={provider_config.saml_configuration_id}, slug='{saml_configuration_slug}') "
+                        f"[INFO] {provider_info} "
+                        f"SAML config (id={config_id}, slug='{saml_configuration_slug}') "
                         "does not match the provider's slug."
```
1. Edit to make more readable.
```suggestion
                        f"[INFO] {provider_info} has "
                        f"SAML config (id={config_id}, slug='{saml_configuration_slug}') "
                        "that does not match the provider's slug."
```
2. Add `DISABLED ` before SAML config if it is disabled.

ktyagiapphelix2u replied:
resolved the info messages

robrap reviewed (commented):
I didn't go through tests yet. This is a start. Thank you.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
Merge branch 'master' into ktyagi/saml3

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,53 +109,73 @@ def _check_provider_configurations(self):
+
+            # Check if provider is disabled
+            provider_disabled = not provider_config.enabled
+            disabled_status = ", enabled=False" if provider_disabled else ""
```
```suggestion
            disabled_status = "enabled=False" if provider_disabled else ""
```

jcapphelix replied:
Comma can be added at line 120.

robrap replied:
[inform] Was restored to original so that we don't see a comma when it is not disabled. Marking resolved.

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -177,6 +199,39 @@ def _check_provider_configurations(self):
+            default_config = SAMLConfiguration.current(provider_config.site_id, 'default')
+            if not default_config or default_config.id is None:
+                # Resolution: Create a SAML configuration for this provider
+                # or create a default configuration for the site
```
```suggestion
                # Resolution: Create/Link a SAML configuration for this provider
                # or create/link a default configuration for the site
```

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -177,6 +199,39 @@ def _check_provider_configurations(self):
+
+            if not default_config.enabled:
+                # Resolution: Enable the default SAML configuration
+                # or create a specific configuration for this provider
```
```suggestion
                # Resolution: Enable the provider's linked SAML configuration
                # or create/link a specific configuration for this provider
```

@ktyagiapphelix2u  only implement this if you think this is more accurate. Ignore if suggestions are not accurate.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -348,24 +349,28 @@ def test_run_checks_outdated_configs(self, mock_set_custom_attribute):
-            mock.call('saml_management_command.total_requiring_attention', 2),
+            mock.call('saml_management_command.total_requiring_attention', 2),  # 1 outdated + 1 null
         ]
         mock_set_custom_attribute.assert_has_calls(expected_calls, any_order=False)
```
I don't have a great suggestion, but there is just too much redundancy across these tests. I'm not sure if helper methods with defaults will help or not.

ktyagiapphelix2u replied:
I've removed all the observability calls (set_custom_attribute) from both the management command and the tests as suggested.

each test was repeating ~10 lines of observability assertions that are no longer needed.

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -192,14 +247,24 @@ def _report_check_summary(self, metrics):
+        self.stdout.write("Informational only:")
+        self.stdout.write(f"  Slug mismatches: {metrics['slug_mismatch_count']['count']}")
+        if metrics['null_config_count']['count'] == 0:
+            self.stdout.write(f"  Missing configs: {metrics['null_config_count']['count']}")
```
What is this for? Why the complexity around "missing configs" that isn't needed for any of the others?

ktyagiapphelix2u replied:
Missing configs: 0" ONLY when count was 0 (under "Informational only")
Showing "Missing configs: X" ONLY when count was > 0 (under "Issues requiring attention")

But i have updated them as missing config just for informational things only

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,53 +109,73 @@ def _check_provider_configurations(self):
-                f"slug={provider_config.slug}, site_id={provider_config.site_id})"
+                f"Provider (id={provider_config.id}, "
+                f"name={provider_config.name}, slug={provider_config.slug}, "
+                f"site_id={provider_config.site_id}{', ' + disabled_status if provider_disabled else ''})"
```
Why did you move the `', '` here, rather than above in `disabled_status` where it was? Then it won't appear if not displaying for disabled provider.

ktyagiapphelix2u replied:
https://github.com/openedx/edx-platform/pull/37377#discussion_r2394214616

I have changed to print the check summary output as it is just for better thing 

the comma and space are part of disabled_status, so they only appear when the provider is actually disabled

robrap reviewed (commented):
Some comments to help progress. I'm not sure what to do because we are clearly spinning on this PR, and there is a lot of code (especially test code) for something that we have lived without. On the one hand, I'd like to land this without putting in too much time. On the other hand, I don't want to dump in a bunch of a new code that needs to be maintained forever that isn't serving a real benefit.

It may make sense to simple drop the observability calls to simplify the code and tests a little. The point is that someone will run the command and you'll just use the output that gets reported. There will still be redundancy in the tests that could possibly get cleaned up, but this would remove some of it.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -348,29 +347,21 @@ def test_run_checks_outdated_configs(self, mock_set_custom_attribute):
-            output
+        outdated_msg = (
+            f'has outdated SAML config (id={old_config.id}) which should be updated to '
+            f'the current SAML config (id={new_config.id})'
```
Since there are multiple providers and messages for some of these tests, the `[WARNING]` could be for the wrong message. Could we simply look for the full message? I know breaking it up might make it more resilient to message changes, but it also makes it more brittle in other ways and harder to understand. 

This same comment applies to all the tests for the full WARNING/INFO message. I _think_ it will be easier to follow.

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -348,29 +347,21 @@ def test_run_checks_outdated_configs(self, mock_set_custom_attribute):
+        self.assertIn('Missing configs: 1', output)  # 1 null from setUp
+        self.assertIn('Issues requiring attention:', output)
         self.assertIn('Outdated: 1', output)
+        self.assertIn('Total issues requiring attention: 1', output)  # 1 outdated only
```
I find most of these assertions distracting from what we are actually testing. Can we strip this down to the following?:
```suggestion
        self.assertIn('Outdated: 1', output)
        self.assertIn('Total issues requiring attention: 1', output)
```

Please do something similar to the other tests, so it is more clear what each test is for.

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +401,148 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+
+        # Verify no Resolution: text appears
+        self.assertNotIn('Resolution:', output)
+
```
Also remove this:
```suggestion
```

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,131 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+            f'Provider (id={disabled_provider.id}, name={disabled_provider.name}, '
+            f'slug={disabled_provider.slug}, site_id={disabled_provider.site_id}, enabled=False)'
+        )
+        self.assertIn(expected_disabled_provider, output)
```
This can be removed ? 

As we are now checking the full message anyways

ktyagiapphelix2u replied:
Removed

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,131 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+            f'has no direct SAML configuration and no matching default configuration was found.'
+        )
+        self.assertIn(expected_warning, output)
+        self.assertIn('Missing configs: 2', output)  # 1 from test + 1 from setUp
```
Comment can be removed.

robrap replied:
[inform] I did ask this comment to be restored, because it isn't obvious without the comment.

jcapphelix commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,131 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+            saml_configuration=None
+        )
+
+        # Mock config with id=None (simulates broken default config)
```
Comment can be removed, code is self-explanatory with vars being `mock_config`

robrap replied:
Agreed. But again, I'm wondering why we need to mock these results. Is it not possible to create test data fails the test?

ktyagiapphelix2u committed:
fix: marketing site configurations

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -107,53 +102,73 @@ def _check_provider_configurations(self):
+                    self.stdout.write(
+                        f"[WARNING] {provider_info} "
+                        f"has SAML config (id={provider_config.saml_configuration_id}, enabled=False)."
+                    )
```
Is there supposed to be a count associated with this?

ktyagiapphelix2u replied:
I added the missing count tracking

robrap commented on common/djangoapps/third_party_auth/management/commands/saml.py:
```diff
@@ -165,41 +180,63 @@ def _check_provider_configurations(self):
+                "no matching default configuration was found."
+            )
+            null_config_count += 1
+            return null_config_count
```
I guess we can just go with the final return in this method.
```suggestion
```

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,122 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
         )
 
+        with mock.patch('common.djangoapps.third_party_auth.models.SAMLConfiguration.current',
+                        return_value=None):
```
Why is this something that needs to be mocked? This is not a case we can get to with regular test data?

ktyagiapphelix2u replied:
Got it i have fixed that

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,122 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+            f'has no direct SAML configuration and no matching default configuration was found.'
+        )
+        self.assertIn(expected_warning, output)
+        self.assertIn('Missing configs: 2', output)
```
It may be useful to add (or restore) a quick comment about why this is 2 and not 1.

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,122 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+            f'has no direct SAML configuration and no matching default configuration was found.'
+        )
+        self.assertIn(expected_warning, output)
+        self.assertIn('Missing configs: 2', output)
```
Same note about the 2. You'll want a short comment explaining.

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -416,57 +392,122 @@ def test_run_checks_slug_mismatches(self, mock_set_custom_attribute):
+
+        output = self._run_checks_command()
+
+        self.assertIn('Missing configs: 1', output)
```
Assuming the following is accurate, you could add this test higher up in the file and then
you could note in comments where this goes from 1 to 2, that the setup data accounts for 1.
```
def test_run_checks_setup_test_data(self):
        """
        Test the --run-checks command against initial setup test data.
        """
        output = self._run_checks_command()

        # The setup data includes a saml provider with a missing config
        self.assertIn('Missing configs: 1', output)
```

**UPDATE:** This is optional. It could even check the warning. Or, you could leave this out if you appropriately comment the other 2s.

ktyagiapphelix2u replied:
Done! Added test_run_checks_setup_test_data to validate the baseline setup data. Updated all test assertions with clarifying comments - the setUp contributes 1 to "Disabled configs" (not "Missing configs" as I initially thought). All tests now pass and clearly show how counts change from the baseline.

robrap reviewed (commented):
Even though there are a bunch of comments, I think this is close. Thanks. The tests don't seem overly dense and redundant any longer.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -412,51 +438,60 @@ def test_run_checks_slug_mismatches(self):
         )
         self.assertIn(expected_warning, output)
-        self.assertIn('Missing configs: 2', output)
+        self.assertIn('Missing configs: 1', output)  # This test's provider with no config
```
test's => tests

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -412,51 +438,60 @@ def test_run_checks_slug_mismatches(self):
         )
         self.assertIn(expected_warning, output)
-        self.assertIn('Missing configs: 2', output)
+        self.assertIn('Missing configs: 1', output)  # This test's provider with disabled default config
```
test's => tests

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -412,51 +438,60 @@ def test_run_checks_slug_mismatches(self):
         self.assertIn(expected_warning, output)
-        self.assertIn('Missing configs: 2', output)
+        self.assertIn('Missing configs: 1', output)  # This test's provider with disabled default config
+        self.assertIn('Disabled configs: 1', output)  # From setUp
```
In theory you could consider this a missing config or a disabled config. I think calling it a disabled config would call better attention to the issue, so I would use the disabled config count for this case, rather than missing config.

ktyagiapphelix2u replied:
Added the message

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -476,7 +511,8 @@ def test_run_checks_with_default_config(self):
         output = self._run_checks_command()
 
-        self.assertIn('Missing configs: 1', output)
+        self.assertIn('Missing configs: 0', output)  # This test's provider has valid default config
```
test's => tests

ktyagiapphelix2u replied:
Corrected

robrap reviewed (commented):
Thanks. Minor comments.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -490,8 +490,8 @@ def test_run_checks_null_config_id(self):
-        self.assertIn('Missing configs: 1', output)  # This test's provider with disabled default config
-        self.assertIn('Disabled configs: 1', output)  # From setUp
+        self.assertIn('Missing configs: 0', output)  # No missing configs since default config exists
+        self.assertIn('Disabled configs: 2', output)  # From setUp + this tests provider with disabled default config
```
I'd put the test first in the comment. Something like:
```
# 1 from this test + 1 from setUp data
```

ktyagiapphelix2u replied:
Resolved

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

robrap commented on common/djangoapps/third_party_auth/management/commands/tests/test_saml.py:
```diff
@@ -490,8 +490,8 @@ def test_run_checks_null_config_id(self):
         self.assertIn(expected_warning, output)
-        self.assertIn('Missing configs: 0', output)  # No missing configs since default config exists
-        self.assertIn('Disabled configs: 2', output)  # From setUp + this tests provider with disabled default config
+        self.assertIn('Missing configs: 1', output)  # 1 from this test (provider with disabled default config)
```
This might require a code change, but my earlier comment was that a provider with a disabled default should be counted as a disabled config, and not a missing config, right? It looks like that was lost.

ktyagiapphelix2u replied:
@robrap right about the logic! The confusion might be because there are two different test scene:

test_run_checks_null_configurations - Provider with no config AND no default available Missing config
test_run_checks_null_config_id - Provider with no direct config but disabled default config Disabled config
We modified the _check_no_config method in saml.py to properly categorize the second case as a disabled config rather than missing config

providers with disabled default configs are now properly categorized as disabled configs rather than missing configs.

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
fix: Improve SAML configuration checks and update warning messages

ktyagiapphelix2u committed:
Merge branch 'master' into ktyagi/saml3

robrap committed:
Merge branch 'master' into ktyagi/saml3
